zebra-chain/src/notes/sapling/nullifiers.rs

```diff
@@ -0,0 +1,62 @@
+#![allow(clippy::unit_arg)]
+#![allow(dead_code)]
+
+use std::io;
+
+use crate::{
+    commitments::sapling::{mixing_pedersen_hash, NoteCommitment},
+    keys::sapling::NullifierDerivingKey,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    treestate::note_commitment_tree::Position,
+};
+
+/// Invokes Blake2s-256 as PRF^nfSapling to derive the nullifier for a
+/// Sapling note.
+///
+/// PRF^nfSapling(ρ*) := BLAKE2s-256("Zcash_nf", nk* || ρ*)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+fn prf_nf(nk: [u8; 32], rho: [u8; 32]) -> [u8; 32] {
+    let hash = blake2s_simd::Params::new()
+        .hash_length(32)
+        .personal(b"Zcash_nf")
+        .to_state()
+        .update(&nk[..])
+        .update(&rho[..])
+        .finalize();
+
+    *hash.as_array()
+}
+
+/// A Nullifier for Sapling transactions
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Nullifier([u8; 32]);
+
+impl From<[u8; 32]> for Nullifier {
+    fn from(buf: [u8; 32]) -> Self {
+        Self(buf)
+    }
+}
+
+impl From<(NoteCommitment, Position, NullifierDerivingKey)> for Nullifier {
```

Comment:
Since From takes ownership of the converted value, this will take ownership of the NullifierDerivingKey. Is that what we want?

Reply:
Nah, I'll fix

Reply:
I seem to have done similar in several places in https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/keys/sapling.rs#L558 and it worked out ok?

Reply:
It seems like that code will have the same problem, right? The issue isn't in the `into()` calls on that line, but for the caller.  For instance, if I have a borrowed NullifierDerivingKey or AuthorizingKey (because I'm not copying the key material everywhere), there's no way to compute the nullifier or the ivk.

Reply:
Like this? https://github.com/ZcashFoundation/zebra/pull/596/commits/4540b57883ab84d1f7c6437b20b74ad155b14def

Reply:
Yes, almost -- except that that's defining it as a borrow of a tuple, but it needs to be a tuple of borrows, because otherwise it's still constraining the storage of the key material.  It's pretty awkward.

Reply:
https://github.com/ZcashFoundation/zebra/pull/596/commits/2c3700c983a784bbcb911aa1d45a74f9f69512ea
?